// @apollosproject/server-core/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createGlobalId", {
  enumerable: true,
  get: function () {
    return Node.createGlobalId;
  }
});
Object.defineProperty(exports, "parseGlobalId", {
  enumerable: true,
  get: function () {
    return Node.parseGlobalId;
  }
});
Object.defineProperty(exports, "withEdgePagination", {
  enumerable: true,
  get: function () {
    return _utils.withEdgePagination;
  }
});
exports.createApolloServerConfig = exports.createContext = exports.createDataSources = exports.createResolvers = exports.createSchema = void 0;

var _lodash = require("lodash");

var _graphqlTag = _interopRequireDefault(require("graphql-tag"));

var Node = _interopRequireWildcard(require("./node"));

var Pagination = _interopRequireWildcard(require("./pagination"));

var Media = _interopRequireWildcard(require("./media"));

var _utils = require("./pagination/utils");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Types that all apollos-church servers will use.
const builtInData = {
  Node,
  Pagination,
  Media
};

const createSchema = data => [_graphqlTag.default`
    type Query {
      _placeholder: Boolean # needed, empty schema defs aren't supported
    }

    type Mutation {
      _placeholder: Boolean # needed, empty schema defs aren't supported
    }
  `, ...(0, _lodash.compact)((0, _lodash.values)({ ...builtInData,
  ...data
}).map(datum => datum.schema))];

exports.createSchema = createSchema;

const createResolvers = data => (0, _lodash.merge)(...(0, _lodash.compact)((0, _lodash.values)({ ...builtInData,
  ...data
}).map(datum => datum.resolver)));

exports.createResolvers = createResolvers;

const getDataSources = data => (0, _lodash.mapValues)({ ...builtInData,
  ...data
}, datum => datum.dataSource);

const getModels = data => (0, _lodash.mapValues)({ ...builtInData,
  ...data
}, datum => datum.model);

const getContextMiddlewares = data => (0, _lodash.compact)((0, _lodash.values)({ ...builtInData,
  ...data
}).map(datum => datum.contextMiddleware));

const createDataSources = data => {
  const dataSources = getDataSources(data);
  return () => {
    const sources = {};
    Object.keys(dataSources).forEach(dataSourceName => {
      if (dataSources[dataSourceName]) {
        sources[dataSourceName] = new dataSources[dataSourceName]();
      }
    });
    return sources;
  };
};

exports.createDataSources = createDataSources;

const createContext = data => ({
  req = {}
} = {}) => {
  const initiatedModels = {}; // For all non-datasource connectors. Right now only `Node`.

  const models = getModels(data);
  let context = {
    models: initiatedModels
  };
  Object.keys(models).forEach(modelName => {
    if (models[modelName]) {
      initiatedModels[modelName] = new models[modelName](context);
    }
  });
  const contextMiddleware = getContextMiddlewares(data);
  contextMiddleware.forEach(middleware => {
    context = middleware({
      req,
      context
    });
  });
  return context;
};

exports.createContext = createContext;

const createApolloServerConfig = data => {
  const context = createContext(data);
  const dataSources = createDataSources(data);
  const schema = createSchema(data);
  const resolvers = createResolvers(data);
  return {
    context,
    dataSources,
    schema,
    resolvers
  };
};

exports.createApolloServerConfig = createApolloServerConfig;
