// @apollosproject/data-connector-rock-auth/lib/data-source.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _apolloServer = require("apollo-server");

var _apolloServerEnv = require("apollo-server-env");

var _moment = _interopRequireDefault(require("moment"));

var _rockApolloDataSource = _interopRequireDefault(require("@apollosproject/rock-apollo-data-source"));

var _token = require("./token");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class AuthDataSource extends _rockApolloDataSource.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "resource", 'Auth');

    _defineProperty(this, "rockCookie", null);

    _defineProperty(this, "userToken", null);

    _defineProperty(this, "getCurrentPerson", async ({
      cookie
    } = {
      cookie: null
    }) => {
      const {
        rockCookie
      } = this.context;
      const userCookie = cookie || rockCookie;

      if (userCookie) {
        const request = await this.request('People/GetCurrentPerson').get({
          options: {
            headers: {
              cookie: userCookie
            }
          }
        });
        return request;
      }

      throw new _apolloServer.AuthenticationError('Must be logged in');
    });

    _defineProperty(this, "fetchUserCookie", async (Username, Password) => {
      try {
        // We use `new Response` rather than string/options b/c if conforms more closely with ApolloRESTDataSource
        // (makes mocking in tests WAY easier to use `new Request` as an input in both places)
        const response = await (0, _apolloServerEnv.fetch)(new _apolloServerEnv.Request(`${this.baseURL}/Auth/Login`, {
          method: 'POST',
          body: JSON.stringify({
            Username,
            Password
          }),
          headers: {
            'Content-Type': 'Application/Json'
          }
        }));
        if (response.status >= 400) throw new _apolloServer.AuthenticationError();
        const cookie = response.headers.get('set-cookie');
        return cookie;
      } catch (err) {
        throw new _apolloServer.AuthenticationError('Invalid Credentials');
      }
    });

    _defineProperty(this, "createSession", async ({
      cookie
    }) => {
      const currentUser = await this.getCurrentPerson({
        cookie
      });
      return this.post('/InteractionSessions', {
        PersonAliasId: currentUser.primaryAliasId
      });
    });

    _defineProperty(this, "authenticate", async ({
      identity,
      password
    }) => {
      try {
        const cookie = await this.fetchUserCookie(identity, password);
        const sessionId = await this.createSession({
          cookie
        });
        const token = (0, _token.generateToken)({
          cookie,
          sessionId
        });
        const {
          userToken,
          rockCookie
        } = (0, _token.registerToken)(token);
        this.context.rockCookie = rockCookie;
        this.context.userToken = userToken;
        this.context.sessionId = sessionId;
        return {
          token,
          rockCookie
        };
      } catch (e) {
        throw e;
      }
    });

    _defineProperty(this, "personExists", async ({
      identity
    }) => {
      const hasUserName = await this.request(`/UserLogins?$filter=UserName eq '${identity}'`).get();

      if (hasUserName.length) {
        return true;
      }

      return false;
    });

    _defineProperty(this, "createUserProfile", async (props = {}) => {
      try {
        const {
          email
        } = props;
        return await this.post('/People', {
          Email: email,
          IsSystem: false,
          // Required by Rock
          Gender: 0 // Required by Rock

        });
      } catch (err) {
        throw new Error('Unable to create profile!');
      }
    });

    _defineProperty(this, "createUserLogin", async (props = {}) => {
      try {
        const {
          email,
          password,
          personId
        } = props;
        return await this.post('/UserLogins', {
          PersonId: personId,
          EntityTypeId: 27,
          // A default setting we use in Rock-person-creation-flow
          UserName: email,
          PlainTextPassword: password,
          LastLoginDateTime: `${(0, _moment.default)().toISOString()}`
        });
      } catch (err) {
        throw new Error('Unable to create user login!');
      }
    });

    _defineProperty(this, "changePassword", async ({
      password
    }) => {
      const currentUser = await this.getCurrentPerson();
      const {
        email,
        id
      } = currentUser;
      const logins = await this.request('/UserLogins').filter(`UserName eq '${email}'`).get();

      if (logins.length > 0) {
        await this.delete(`/UserLogins/${logins[0].id}`);
      }

      await this.createUserLogin({
        personId: id,
        email,
        password
      });
      return this.authenticate({
        identity: email,
        password
      });
    });

    _defineProperty(this, "registerPerson", async ({
      email,
      password
    }) => {
      const personExists = await this.personExists({
        identity: email
      });
      if (personExists) throw new Error('User already exists!');
      const personId = await this.createUserProfile({
        email
      });
      await this.createUserLogin({
        email,
        password,
        personId
      });
      const token = await this.authenticate({
        identity: email,
        password
      });
      return token;
    });
  }

}

exports.default = AuthDataSource;
