// @apollosproject/data-connector-rock-auth/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "registerToken", {
  enumerable: true,
  get: function () {
    return _token.registerToken;
  }
});
Object.defineProperty(exports, "generateToken", {
  enumerable: true,
  get: function () {
    return _token.generateToken;
  }
});
Object.defineProperty(exports, "schema", {
  enumerable: true,
  get: function () {
    return _dataSchema.authSchema;
  }
});
Object.defineProperty(exports, "dataSource", {
  enumerable: true,
  get: function () {
    return _dataSource.default;
  }
});
exports.resolver = exports.contextMiddleware = void 0;

var _serverCore = require("@apollosproject/server-core");

var _lodash = require("lodash");

var _token = require("./token");

var _dataSchema = require("@apollosproject/data-schema");

var _dataSource = _interopRequireDefault(require("./data-source"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const contextMiddleware = ({
  req,
  context: ctx
}) => {
  if ((0, _lodash.get)(req, 'headers.authorization')) {
    const {
      userToken,
      rockCookie,
      sessionId
    } = (0, _token.registerToken)(req.headers.authorization);

    if (sessionId) {
      return { ...ctx,
        userToken,
        rockCookie,
        sessionId
      };
    }
  }

  return ctx;
};

exports.contextMiddleware = contextMiddleware;
const resolver = {
  Query: {
    currentUser: (root, args, {
      dataSources
    }) => dataSources.Auth.getCurrentPerson()
  },
  AuthenticatedUser: {
    id: ({
      id
    }, args, context, {
      parentType
    }) => (0, _serverCore.createGlobalId)(id, parentType.name),
    profile: authUser => authUser
  },
  Authentication: {
    user: (root, args, {
      dataSources
    }) => dataSources.Auth.getCurrentPerson()
  },
  Mutation: {
    authenticate: (root, {
      identity,
      password
    }, {
      dataSources
    }) => dataSources.Auth.authenticate({
      identity,
      password
    }),
    changePassword: (root, {
      password
    }, {
      dataSources
    }) => dataSources.Auth.changePassword({
      password
    }),
    registerPerson: (root, args, {
      dataSources
    }) => dataSources.Auth.registerPerson(args)
  }
};
exports.resolver = resolver;
