// @apollosproject/data-connector-rock-auth/lib/token.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateToken = exports.registerToken = exports.parseToken = void 0;

var _apolloServer = require("apollo-server");

var _jsonwebtoken = _interopRequireDefault(require("jsonwebtoken"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const secret = process.env.SECRET || 'ASea$2gadj#asd0';

const parseToken = token => _jsonwebtoken.default.verify(token, secret);

exports.parseToken = parseToken;

const registerToken = token => {
  try {
    const {
      cookie,
      sessionId
    } = parseToken(token);
    return {
      userToken: token,
      rockCookie: cookie,
      sessionId
    };
  } catch (e) {
    throw new _apolloServer.AuthenticationError('Invalid token');
  }
};

exports.registerToken = registerToken;

const generateToken = params => _jsonwebtoken.default.sign({ ...params
}, secret, {
  expiresIn: '60d'
});

exports.generateToken = generateToken;
