// @apollosproject/config/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _jsYaml = _interopRequireDefault(require("js-yaml"));

var _attachEnvVariables = _interopRequireDefault(require("./attach-env-variables"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Config {
  constructor() {
    _defineProperty(this, "config", {
      ANALYTICS: {},
      CLOUDINARY: {},
      ROCK: {},
      ROCK_CONSTANTS: {},
      ROCK_MAPPINGS: {},
      BIBLE_API: {}
    });

    this.attachConfigToClass(this.config);
  }

  loadYaml({
    configPath
  }) {
    if (!configPath) {
      throw new Error('Config must be specifed in ApollosConfig');
    }

    let file;

    try {
      file = _fs.default.readFileSync(configPath, 'utf8');
    } catch (e) {
      throw new Error(`${configPath} does not exist`);
    }

    this._yml = _jsYaml.default.safeLoad(file);
    this.config = Object.assign(this.config, (0, _attachEnvVariables.default)(this._yml));
    this.attachConfigToClass(this.config);
    return this;
  }

  loadJs(data = {}) {
    Object.keys(data).forEach(key => {
      if (this.config[key]) {
        // Deep merge defaults if they exist.
        this.config[key] = Object.assign(this.config[key], data[key]);
      } else {
        // Otherwise attach results to config.
        this.config[key] = data[key];
      }
    });
    this.attachConfigToClass(this.config);
  } // Allows you to do Config.PROPERTY


  attachConfigToClass(config) {
    Object.keys(config).forEach(key => {
      Object.defineProperty(this, key, {
        enumerable: true,
        configurable: true,
        writable: true,
        value: config[key]
      });
    });
  }

}

const config = new Config();
var _default = config;
exports.default = _default;
