// @apollosproject/rock-apollo-data-source/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _apolloDatasourceRest = require("apollo-datasource-rest");

var _config = _interopRequireDefault(require("@apollosproject/config"));

var _lodash = require("lodash");

var _apolloServerEnv = require("apollo-server-env");

var _cursor = require("./cursor");

var _requestBuilder = _interopRequireDefault(require("./request-builder"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const {
  ROCK
} = _config.default;

class RockApolloDataSource extends _apolloDatasourceRest.RESTDataSource {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "expanded", false);

    _defineProperty(this, "baseURL", ROCK.API_URL);

    _defineProperty(this, "rockToken", ROCK.API_TOKEN);

    _defineProperty(this, "nodeFetch", _apolloServerEnv.fetch);

    _defineProperty(this, "normalize", data => {
      if (Array.isArray(data)) return data.map(this.normalize);
      if (typeof data !== 'object') return data;
      const normalizedValues = (0, _lodash.mapValues)(data, this.normalize);
      return (0, _lodash.mapKeys)(normalizedValues, (value, key) => (0, _lodash.camelCase)(key));
    });
  }

  didReceiveResponse(response, request) {
    // Can't use await b/c of `super` keyword
    return super.didReceiveResponse(response, request).then(parsedResponse => this.normalize(parsedResponse));
  }

  willSendRequest(request) {
    request.headers.set('Authorization-Token', this.rockToken);
    request.headers.set('user-agent', 'Apollos');
    request.headers.set('Content-Type', 'application/json');
  }

  request(resource = this.resource) {
    return new _requestBuilder.default({
      resource,
      connector: this,
      defaultOptions: this.expanded ? {
        loadAttributes: 'expanded'
      } : null
    });
  }

  async paginate({
    cursor,
    args: {
      after,
      first = 20
    } = {}
  }) {
    let skip = 0;

    if (after) {
      const parsed = (0, _cursor.parseCursor)(after);

      if (parsed && Object.hasOwnProperty.call(parsed, 'position')) {
        skip = parsed.position + 1;
      } else {
        throw new Error(`An invalid 'after' cursor was provided: ${after}`);
      }
    }

    const edges = cursor ? cursor.top(first).skip(skip).transform(result => result.map((node, i) => ({
      node,
      cursor: (0, _cursor.createCursor)({
        position: i + skip
      })
    }))).get() : [];
    return {
      edges
    };
  }

}

exports.default = RockApolloDataSource;
