// @apollosproject/rock-apollo-data-source/lib/request-builder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _withQuery = _interopRequireDefault(require("with-query"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// Simple request builder for querying the Rock API.
// Would probably work against most OData APIs, but built to just
// tackle the specific needs of Apollos on top of Rock.
class RockRequestBuilder {
  constructor({
    connector,
    resource,
    defaultOptions = null
  }) {
    _defineProperty(this, "query", {});

    _defineProperty(this, "transforms", []);

    _defineProperty(this, "options", {});

    _defineProperty(this, "get", ({
      options = {},
      body = {}
    } = {}) => this.connector.get(this.path, body, { ...options,
      ...this.options
    }).then(results => {
      if (this.transforms.length) return this.transforms.reduce((current, transformer) => transformer(current), results);
      return results;
    }));

    _defineProperty(this, "find", id => {
      this.resourceId = id;
      return this;
    });

    _defineProperty(this, "filter", filter => {
      const key = '$filter';

      if (this.query[key]) {
        this.query[key] = `(${this.query[key]}) or (${filter})`;
      } else {
        this.query[key] = filter;
      }

      return this;
    });

    _defineProperty(this, "cache", ({
      ttl
    }) => {
      this.options.ttl = ttl;
      return this;
    });

    _defineProperty(this, "expand", expand => {
      let {
        $expand
      } = this.query;

      if (!$expand) {
        $expand = [];
      } else {
        $expand = $expand.split(',');
      }

      $expand.push(expand);
      this.query.$expand = $expand.join(',');
      return this;
    });

    _defineProperty(this, "orderBy", (name, direction = 'asc') => {
      this.query.$orderby = `${name} ${direction}`;
      return this;
    });

    _defineProperty(this, "top", top => {
      this.query.$top = top;
      return this;
    });

    _defineProperty(this, "skip", skip => {
      this.query.$skip = skip;
      return this;
    });

    _defineProperty(this, "select", select => {
      this.query.$select = select;
      return this;
    });

    _defineProperty(this, "transform", func => {
      this.transforms.push(func);
      return this;
    });

    this.connector = connector;
    this.resource = resource;

    if (defaultOptions) {
      this.query = defaultOptions;
    }
  }

  get path() {
    let path = [this.resource];
    if (this.resourceId) path.push(this.resourceId);
    path = path.join('/');
    path = (0, _withQuery.default)(path, this.query);
    return path;
  }
  /**
   * Sends a GET request to the server, resolves with results
   * @returns promise
   */


}

exports.default = RockRequestBuilder;
