// @apollosproject/data-connector-cloudinary/lib/cloudinary.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.config = config;
exports.default = withCloudinary;
Object.defineProperty(exports, "cloudinary", {
  enumerable: true,
  get: function () {
    return _cloudinary.default;
  }
});

var _cloudinary = _interopRequireDefault(require("cloudinary"));

var _config = _interopRequireDefault(require("@apollosproject/config"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const {
  CLOUDINARY
} = _config.default; // Cloudinary picks up settings from `env`,
// no way to pass straight into config if using `CLOUDINARY_URL format`
// Exposed as Utiliy function to reset config (mostly used in tests)

function config() {
  process.env.CLOUDINARY_URL = CLOUDINARY.URL;

  _cloudinary.default.config(true);

  _cloudinary.default.config({
    private_cdn: false,
    secure: true
  });
} // we always want to setup the config on boot.


config();

const cleanUrl = url => url.replace(/:(443|80)/, '');

function withCloudinary(_url = '', options) {
  const url = cleanUrl(_url); // If we call this function twice, only the first transform will be applied

  if (url.startsWith('https://res.cloudinary.com')) {
    return url;
  }

  if (CLOUDINARY.URL) {
    return _cloudinary.default.url(url, {
      type: 'fetch',
      fetch_format: 'auto',
      width: '1600',
      crop: 'limit',
      ...options
    });
  }

  return url;
}
