// @apollosproject/data-connector-rock-actions/lib/interactions/data-source.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _serverCore = require("@apollosproject/server-core");

var _rockApolloDataSource = _interopRequireDefault(require("@apollosproject/rock-apollo-data-source"));

var _config = _interopRequireDefault(require("@apollosproject/config"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Interactions extends _rockApolloDataSource.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "resource", 'Interactions');
  }

  async createContentItemInteraction({
    itemId,
    operationName,
    itemTitle
  }) {
    const {
      dataSources: {
        RockConstants,
        Auth
      }
    } = this.context;
    const {
      id
    } = (0, _serverCore.parseGlobalId)(itemId);
    const interactionComponent = await RockConstants.contentItemInteractionComponent({
      contentItemId: id,
      // Don't want to recreate channels if name changes
      // In the future, we could use itemTitle here.
      contentName: id
    });
    const currentUser = await Auth.getCurrentPerson();
    const interactionId = await this.post('/Interactions', {
      PersonAliasId: currentUser.primaryAliasId,
      InteractionComponentId: interactionComponent.id,
      InteractionSessionId: this.context.sessionId,
      Operation: operationName,
      InteractionDateTime: new Date().toJSON(),
      InteractionSummary: `${operationName} - ${itemTitle}`,
      InteractionData: `${_config.default.APP.DEEP_LINK_HOST}://Interactions/ContentSingle?itemId=${itemId}`
    });
    return this.get(`/Interactions/${interactionId}`);
  }

}

exports.default = Interactions;
