// @apollosproject/data-connector-rock-content/lib/content-channels/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "dataSource", {
  enumerable: true,
  get: function () {
    return _dataSource.default;
  }
});
Object.defineProperty(exports, "schema", {
  enumerable: true,
  get: function () {
    return _dataSchema.contentChannelSchema;
  }
});
exports.resolver = void 0;

var _serverCore = require("@apollosproject/server-core");

var _dataSource = _interopRequireDefault(require("./data-source"));

var _dataSchema = require("@apollosproject/data-schema");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// export { default as model } from './model';
const resolver = {
  Query: {
    contentChannels: (root, args, context) => context.dataSources.ContentChannel.getRootChannels()
  },
  ContentChannel: {
    id: ({
      id
    }, args, context, {
      parentType
    }) => (0, _serverCore.createGlobalId)(id, parentType.name),
    childContentItemsConnection: ({
      id
    }, args, {
      dataSources
    }) => dataSources.ContentItem.paginate({
      cursor: dataSources.ContentItem.byContentChannelId(id),
      args
    }),
    iconName: () => 'text' // TODO

  }
};
exports.resolver = resolver;
