// @apollosproject/server-core/lib/testUtils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createTestHelpers = void 0;

var _apolloServer = require("apollo-server");

var _dataSchema = require("@apollosproject/data-schema");

var _apolloServerCaching = require("apollo-server-caching");

var _ = require("..");

/* eslint-disable import/prefer-default-export */
const createTestHelpers = models => {
  const serverConfig = (0, _.createApolloServerConfig)(models);

  const getContext = req => {
    const testContext = serverConfig.context(req);
    const testDataSources = serverConfig.dataSources(); // Apollo Server does this internally.

    Object.values(testDataSources).forEach(dataSource => {
      if (dataSource.initialize) {
        dataSource.initialize({
          context: testContext,
          cache: _apolloServerCaching.KeyValueCache
        });
      }
    });
    testContext.dataSources = testDataSources;
    return testContext;
  };

  const getSchema = (schemas = []) => (0, _apolloServer.makeExecutableSchema)({
    typeDefs: [...serverConfig.schema, ...schemas, _dataSchema.testSchema],
    resolvers: serverConfig.resolvers,
    resolverValidationOptions: {
      requireResolversForResolveType: false
    }
  });

  return {
    getSchema,
    getContext
  };
};

exports.createTestHelpers = createTestHelpers;
