// @apollosproject/data-connector-people/lib/people/data-source.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _apolloServer = require("apollo-server");

var _formData = _interopRequireDefault(require("form-data"));

var _lodash = require("lodash");

var _rockApolloDataSource = _interopRequireDefault(require("@apollosproject/rock-apollo-data-source"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Person extends _rockApolloDataSource.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "resource", 'People');

    _defineProperty(this, "getFromId", id => this.request().find(id).get());

    _defineProperty(this, "getFromEmail", email => this.request().filter(`Email eq '${email}'`).get());

    _defineProperty(this, "updateProfile", async fields => {
      const currentPerson = await this.context.dataSources.Auth.getCurrentPerson();
      if (!currentPerson) throw new _apolloServer.AuthenticationError('Invalid Credentials');
      const fieldsAsObject = fields.reduce((accum, {
        field,
        value
      }) => ({ ...accum,
        [field]: value
      }), {});
      await this.patch(`/People/${currentPerson.id}`, fieldsAsObject);
      return { ...currentPerson,
        ...(0, _lodash.mapKeys)(fieldsAsObject, (_, key) => (0, _lodash.camelCase)(key))
      };
    });

    _defineProperty(this, "uploadProfileImage", async (file, length) => {
      const {
        stream,
        filename
      } = await file;
      const data = new _formData.default();
      data.append('file', stream, {
        filename,
        knownLength: length
      });
      const response = await this.nodeFetch(`${this.baseURL}/BinaryFiles/Upload?binaryFileTypeId=5`, {
        method: 'POST',
        body: data,
        headers: {
          'Authorization-Token': this.rockToken,
          ...data.getHeaders()
        }
      });
      const photoId = await response.text();
      return this.updateProfile([{
        field: 'PhotoId',
        value: photoId
      }]);
    });
  }

}

exports.default = Person;
