// @apollosproject/data-connector-people/lib/people/resolver.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _serverCore = require("@apollosproject/server-core");

var _default = {
  Query: {
    people: (root, {
      email
    }, {
      dataSources
    }) => dataSources.Person.getFromEmail(email)
  },
  Mutation: {
    updateProfileField: (root, {
      input: {
        field,
        value
      }
    }, {
      dataSources
    }) => dataSources.Person.updateProfile([{
      field,
      value
    }]),
    updateProfileFields: (root, {
      input
    }, {
      dataSources
    }) => dataSources.Person.updateProfile(input),
    uploadProfileImage: async (root, {
      file,
      size
    }, {
      dataSources
    }) => dataSources.Person.uploadProfileImage(file, size)
  },
  Person: {
    id: ({
      id
    }, args, context, {
      parentType
    }) => (0, _serverCore.createGlobalId)(id, parentType.name),
    photo: ({
      photo: {
        url
      }
    }) => ({
      uri: url
    }),
    firstName: ({
      firstName
    }) => typeof firstName === 'object' ? '' : firstName,
    lastName: ({
      lastName
    }) => typeof lastName === 'object' ? '' : lastName
  }
};
exports.default = _default;
